types/ember__array/proxy.d.ts

```diff
@@ -9,7 +14,9 @@ import NativeArray from "@ember/array/-private/native-array";
  */
 interface ArrayProxy<T> extends MutableArray<T> {}
 declare class ArrayProxy<T> extends EmberObject.extend(MutableArray as {}) {
-    content: NativeArray<T>;
+    // Despite docs only claiming that Ember.Array and Ember.MutableArray are supported, any object that has
+    // `length` and `objectAt` also work along with native Arrays.
```

Comment:
Ahhh, alas: we *shouldn't* support anything that isn't explicitly supported from what the docs say is public API. This is an implementation detail – probably trying to do a cheap version of a "brand check" in this case. I think we'd be okay if we just required it to be `Ember.Array<T> | Ember.MutableArray<T>` if that's what's explicitly supported by the API docs.

Reply:
The internal code check is here:
https://github.com/emberjs/ember.js/blob/master/packages/@ember/-internals/runtime/lib/system/array_proxy.js#L270-L273

Reply:
This is using `isArray` which in its own docs says:

```
  Objects are considered Array-like if any of the following are true:
    - the object is a native Array
    - the object has an objectAt property
    - the object is an Object, and has a length property
```

Reply:
Ah, I see – I misunderstood. 👍🏼 Thanks for clarifying!

Reply:
While on the one hand I generally want to go with the docs, I think given the implementation, the docs are probably wrong. I also think it's likely that, since extending the native Array prototype isn't a default anymore, people are using ArrayProxy with pure arrays in the wild. One potential concern is what happens when the content array is updated when using a native array. I'll investigate this.

Reply:
I guess this is already generally an issue elsewhere. If you're working with non-Ember Arrays, updating their contents won't be recognized.

Reply:
See also https://github.com/emberjs/ember.js/pull/19399